// @2sic.com/dnn-sxc-angular/dist/http/interceptor.js
import { Injectable } from '@angular/core';
import { Context } from "../context/context.service";
var Interceptor = /** @class */ (function () {
    function Interceptor(context) {
        this.context = context;
    }
    Interceptor.prototype.intercept = function (req, next) {
        return this.context.all$.take(1)
            .flatMap(function (ctx) {
            // Clone the request and update the url with 2sxc params.
            var newReq = req.clone({
                url: ctx.sxc.resolveServiceUrl(req.url),
                setHeaders: {
                    ModuleId: ctx.moduleId.toString(),
                    TabId: ctx.tabId.toString(),
                    ContentBlockId: ctx.contentBlockId.toString(),
                    RequestVerificationToken: ctx.antiForgeryToken,
                    'X-Debugging-Hint': 'bootstrapped by Sxc4Angular',
                },
            });
            return next.handle(newReq);
        });
    };
    Interceptor.decorators = [
        { type: Injectable },
    ];
    /** @nocollapse */
    Interceptor.ctorParameters = function () { return [
        { type: Context, },
    ]; };
    return Interceptor;
}());
export { Interceptor };
