// @2sic.com/dnn-sxc-angular/dist/beta/edit.js
import { Context } from '../context/context.service';
import { Directive, ElementRef, Input, } from '@angular/core';
var SxcToolbarDirective = /** @class */ (function () {
    function SxcToolbarDirective(elementRef, context) {
        this.elementRef = elementRef;
        this.context = context;
        this.toolbar = {};
        this.settings = {};
    }
    SxcToolbarDirective.prototype.ngOnInit = function () {
        var _this = this;
        this.context.sxc$.subscribe(function (sxcSimple) {
            var sxc = sxcSimple;
            if (!sxc.manage)
                return; // edit not available, probably not logged in
            _this.setHtml(sxc.manage.getToolbar(_this.toolbar, _this.settings));
        });
    };
    SxcToolbarDirective.prototype.setHtml = function (html) {
        this.elementRef.nativeElement.innerHTML = html;
    };
    SxcToolbarDirective.decorators = [
        { type: Directive, args: [{
                    selector: 'sxc-toolbar'
                },] },
    ];
    /** @nocollapse */
    SxcToolbarDirective.ctorParameters = function () { return [
        { type: ElementRef, },
        { type: Context, },
    ]; };
    SxcToolbarDirective.propDecorators = {
        "entityId": [{ type: Input },],
        "toolbar": [{ type: Input },],
        "settings": [{ type: Input },],
    };
    return SxcToolbarDirective;
}());
export { SxcToolbarDirective };
